### PR TITLE
Revise Expect.within and add Expect.notWithin prior to merging

### PR DESCRIPTION
**Pull request by drathier**
Implemented a better float comparison function. Not perfect, but much better.

### CODE REVIEW

**Comment by rtfeldman** [comment]
I like this a lot! @drathier are there any edge cases you think this doesn't cover well?

Thoughts @mgold

**Comment by rtfeldman** [comment]
@drathier might be good to copy/paste the tests from https://github.com/elm-lang/core/pull/786 into the `elm-community/elm-test` suite, to guard against future regressions on a real-world suite.

File: src/Expect.elm [hunk 1]
```diff
@@ -265,7 +266,45 @@ which argument is which:
...
+withinCompare tolerance a b =
+    let
+        delta =
+            abs (a - b)
+
+        -- largest non-infinite value expressible in a 64bit float
+        float64maxValue =
+            (2 - (2 ^ -52)) * 2 ^ 1023
+
+        -- smallest positive value representable in a 64bit float with a non-zero radix; the smallest normal number
```
Comment (drathier) [inline src/Expect.elm]: Below this constant we start losing precision.

**Comment by drathier** [comment]
@rtfeldman Which tests did you have in mind? The trigonometry tests? Or the basic math tests?

**Comment by mgold** [comment]
Sorry I haven't had a chance to look at this thoroughly until now!

Clearly you know more about floating point arithmetic than I do, so I'm going to give a 👍 after some minor style tweaks.

I'm a little wary of `notWithin` since that wasn't on the original design thread. @rtfeldman you're the champion of keeping APIs small and not guessing at use-cases, thoughts here?

I'd also like to get whatever tests Richard has in mind added before we merge.

**Comment by drathier** [comment]
@mgold Is there a recommended max line length for Elm code? It's nice that you like the tests :) That's the part that I was a bit unsure about; the limits are chosen quite arbitrarily. 

I would like a test to make sure `withinCompare` is commutative. Just randomize two arguments and a tolerance and assert `wc tol a b == wc tol b a`.

**Comment by mgold** [comment]
> What's the alternative way to simulate notWithin using within?

There's no way to invert an expectation (outside of special things we do in testing elm-test itself). So, that's part of an argument for including `notWithin`. I can't think of when you'd need it though, but granted it would be silly to not have `Expect.notEqual` and this is that for floats so... sure.

I don't think you can test expectations for equality, since they will contain different strings if they fail. Let me push up a commit.

**Comment by drathier** [comment]
@mgold That's a feature request if I ever saw one. Right now, I'm trying to write a test that checks that `Expect.notWithin epsilon a a` always fails.

**Comment by mgold** [comment]
> That's a feature request if I ever saw one.

The problem is that passing tests don't carry a message with them, so when you flip it, you don't have anything useful to say.

**Comment by drathier** [comment]
@mgold True, but you could just pass that string as an argument to the `Expect.failure` function, like we do with `Expect.true`. 

Could you add the irreflexivity test that I removed in https://github.com/elm-community/elm-test/pull/93/commits/c74c9e52ed3728395aba4db35b94e62e5b219613 back again? Github + Travis isn't the best dev environment.

**Comment by mgold** [comment]
Drat, either I hit the wrong button or it merged automatically.

**Comment by drathier** [comment]
@mgold wrong button; travis failed 3 builds with out of memory errors, and the 4th is still running.

**Comment by mgold** [comment]
> travis failed 3 builds with out of memory errors

Yeah, well that's because you tried to equate expectations and shrinking blew up :P

**Comment by mgold** [comment]
Let's continue this back on #87. I'll look at adding your irreflexivity tests.
